rbc/irtools.py

```diff
@@ -460,24 +558,246 @@ def cg_fflush(builder):
 @extending.intrinsic
 def fflush(typingctx):
     """fflush that can be called from Numba jit-decorated functions.
+
+    Note: fflush is available only for CPU target.
     """
     sig = nb_types.void(nb_types.void)
 
     def codegen(context, builder, signature, args):
-        cg_fflush(builder)
+        if typingctx.target_info.is_cpu:
+            cg_fflush(builder)
 
     return sig, codegen
 
 
 @extending.intrinsic
 def printf(typingctx, format_type, *args):
     """printf that can be called from Numba jit-decorated functions.
+
+    Note: printf is available only for CPU target.
     """
     if isinstance(format_type, nb_types.StringLiteral):
         sig = nb_types.void(format_type, nb_types.BaseTuple.from_types(args))
 
         def codegen(context, builder, signature, args):
-            cgutils.printf(builder, format_type.literal_value, *args[1:])
-            cg_fflush(builder)
+            if typingctx.target_info.is_cpu:
+                cgutils.printf(builder, format_type.literal_value, *args[1:])
+                cg_fflush(builder)
 
         return sig, codegen
+
+
+def get_llvm_name(f):
+    """Return normalized name of a llvm intrinsic name.
+    """
+    if f.startswith('llvm.'):
+        nf = f[5:]
+        p = nf.rsplit('.', 1)[-1]
+        if p in ['p0i8',
+                 'f64', 'f32',
+                 'i1', 'i8', 'i16', 'i32', 'i64', 'i128']:
+            # TODO: implement more robust suffix dropper
+            nf = nf[:-len(p)-1]
+        return nf
+    return f
+
+
+def get_libdevice_functions():
+    """Return all libdevice function names with prefix `__nv_` removed.
+
+    Reference: https://docs.nvidia.com/cuda/libdevice-users-guide/function-desc.html#function-desc
+    """
```

Comment:
If numba is >= (0, 52), then, this list is available on `numba.cuda.libdevice`. See:
https://github.com/numba/numba/blob/master/numba/cuda/libdevice.py

Reply:
Good to know. Will keep the list here (it is moved to rbc/libfuncs.py) as rbc has its own convention of storing libraries function names (see libfuncs.py).